File: src/main/java/io/github/shaksternano/borgar/media/io/MediaReaderFactory.java
package io.github.shaksternano.borgar.media.io;

import io.github.shaksternano.borgar.media.io.reader.MediaReader;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

public interface MediaReaderFactory<T> {

    MediaReader<T> createReader(File media, String format) throws IOException;

    MediaReader<T> createReader(InputStream media, String format) throws IOException;
}

File: core/src/main/kotlin/media/MediaReaderFactory.kt
package com.shakster.borgar.core.media

import com.shakster.borgar.core.io.DataSource
import com.shakster.borgar.core.media.reader.AudioReader
import com.shakster.borgar.core.media.reader.ImageReader
import com.shakster.borgar.core.media.reader.MediaReader

interface MediaReaderFactory<T : VideoFrame<*>> {

    val supportedFormats: Set<String>

    suspend fun create(input: DataSource): MediaReader<T>
}

interface ImageReaderFactory : MediaReaderFactory<ImageFrame> {
    override suspend fun create(input: DataSource): ImageReader
}

interface AudioReaderFactory : MediaReaderFactory<AudioFrame> {
    override suspend fun create(input: DataSource): AudioReader
}
